mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -1306,4 +1306,33 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
    * satisfies this requirement, we simply use a type alias here.
    */
   private[recommendation] type ALSPartitioner = org.apache.spark.HashPartitioner
+
+  /**
+   * Private function to checkpoint the RDD and clean up its all of its parents' shuffles eagerly.
+   */
+  private[spark] def checkpointAndCleanParents[T](rdd: RDD[T], blocking: Boolean = false): Unit = {
+    val sc = rdd.sparkContext
+    // If there is no reference tracking we skip clean up.
+    if (sc.cleaner.isEmpty) {
+      return rdd.checkpoint()
```

Comment:
If we bail out early here, then we don't do a `count`. So in the explicit case, I believe the checkpoint won't work correctly at [L677](https://github.com/apache/spark/pull/11919/files#diff-be65dd1d6adc53138156641b610fcadaR677).

Perhaps we can refactor like:
```scala
val cleaner = rdd.sparkContext.cleaner
val deps = rdd.dependencies
rdd.checkpoint()
rdd.count()
cleaner.foreach { c =>
  deps.foreach(_ => cleanEagerly(_, c))
}
```